### ISSUE TITLE
Poor Array.shift() Performance

### ISSUE THREAD

**Issue by Lonniebiz** [denoland/deno#5259]
The v8 engine offers poor performance when removing the first item of a large array. I made a test that showcases this:

```
function shiftTest(numberOfElements)
{
	let x = 0;
	let ary = [];
	while (x !== numberOfElements)
	{
		ary.push(x);
		x += 1;
	}
	console.time(`Shifted ${numberOfElements} elements in `);
	while(ary.length > 0)
	{
		ary.shift();
	}
	console.timeEnd(`Shifted ${numberOfElements} elements in `);
}

function runTest()
{
	console.time("All tests combined completed in ");
	let y = 0;
	while (y < 15000)
	{
		shiftTest(y);
		y += 100;
	}
	while (y < 15100)
	{
		shiftTest(y);
		y += 1;
	}
	y = 20000;
	while (y < 150001)
	{
		shiftTest(y);
		y += 5000;
	}
	console.timeEnd("All tests combined completed in ");
}
runTest();
```
For me, this is how long the test takes:
```
23.644s - Node 14.0.0 (V8: 8.1.307.30-node.30)
 9523ms - Deno 1.0.0-rc3 (V8: 8.4.300)
 9186ms - Chromium 81.0.4044.138 (V8: 8.1.307.32)
  375ms - Firefox 76.0
```
I think it is interesting that Deno beats Node.js (taking less than half the time to complete the test), but it is not quite as fast as Chromium 81. The V8 engine could obviously do much better, considering that Firefox accomplishes the task in a fraction of the time V8 takes. If you'd like to run these tests for yourself, here's a [page for testing in the browsers](http://www.lonniebest.com/BadShiftPerformance/).

I've submitted an upstream [bug 1001825](https://bugs.chromium.org/p/chromium/issues/detail?id=1001825) to V8 regarding this issue, but right now its status is "won't fix".

I'm submitting here too, because the issue indirectly effects deno, and secondly, I thought the results would be of interest to the performance-conscious developers of the deno project.

**Comment by Vaatik** [comment]
Nice catch, even though it affects Deno, Deno is not the cause of it thus I wouldn't have made this a PR personally. But since you said you flagged it to V8 I believe it's cool you putted it here. It might just clutter the open issues more tho

**Comment by Lonniebiz** [comment]
@Vaatik : I agree that Deno is not the cause and if you need to close this to eliminate clutter I will not be offended. At the same time, I'm glad that some people from the deno project have gotten to see this performance issue, as I know they care about performance and may ultimately have influence on helping against the V8 bug's current "wont' fix status".

**Comment by hayd** [comment]
In lieu of this being solved in v8, I think you want to use a different datastructure.

e.g. I used to have an implementation of circular array in hayd/deno-collections, which may give better amortized behavior in this case, I'll bring some of these back soon... although this is an esoteric example (perhaps why V8 are labeling it "won't fix" ?)

**Comment by Lonniebiz** [comment]
@hayd : Perhaps V8 considers this an esoteric example too and maybe that's why  the status is [won't fix](https://bugs.chromium.org/p/chromium/issues/detail?id=1001825).

Around the time I submitted that bug, I thought deeply about the problem and came up 3 or 4 different ideas to address it. Some of those ideas were pretty weird. For example, in a class I named "NegativeArray", I "pushed" items into the array starting at index 4294967294, counting down!

In another solution, I didn't actually remove array items every time they were shifted. Instead, I set them equal to undefined, which was super efficient, but on extremely large arrays, I had to also delete indexes (in addition to setting them to undefined) or it would run out of heap memory. Deleting (versus Undefined) is less efficient.

The last idea I tried was object linking. Where, instead of an array, each item was linked to the next item. This was pretty fast, but not as fast as the array where I'd set items to undefined instead of actually shifting them.

If you're interested in those classes I created, I've zipped them up below: 
[FIFO_Ideas.zip](https://github.com/denoland/deno/files/4624116/FIFO_Ideas.zip)

This code was just proof of concept to myself, so you're seeing raw things I never intended to publish.
